File: 287-find-the-duplicate-number/find-the-duplicate-number.go
func findDuplicate(nums []int) int {
    n := len(nums)
    l, r := 1, n-1

    for l < r{
        m := l + (r-l)/2
        cnt := 0

        for _, nm := range nums {
            if nm <= m {
                cnt++
            }
        }

        if cnt > m {
            r = m
        }else {
            l = m + 1
        }
    }
    return l
}

File: 287-find-the-duplicate-number/find-the-duplicate-number.php
class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function findDuplicate($nums) {
        $slow = 0;
        $fast = 0;

        while (1) {
            $slow = $nums[$slow];
            $fast = $nums[$nums[$fast]];
            if ($slow == $fast){
                break;
            } 
        }

        $slow = 0;
        while($slow != $fast) {
            $slow = $nums[$slow];
            $fast = $nums[$fast];
        }

        return $slow;
    }
}